habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -754,6 +754,74 @@ def within(
     return containment_ids
 
 
+def ontop(
+    sim: habitat_sim.Simulator,
+    objectA: Union[
+        habitat_sim.physics.ManagedRigidObject,
+        habitat_sim.physics.ManagedArticulatedObject,
+        int,
+    ],
+    do_collision_detection: bool = True,
+    vertical_normal_error_threshold: float = 0.75,
+) -> List[int]:
+    """
+    Get a list of all object ids or objects that are "ontop" of a particular objectA.
+    Concretely, 'ontop' is defined as: contact points between objectA and objectB have vertical normals "upward" relative to objectA.
+    This function uses collision points to determine which objects are resting on or contacting the surface of objectA.
+
+    :param sim: The Simulator instance.
+    :param objectA: The ManagedRigidObject or object id for which to query the 'ontop' set.
+    :param do_collision_detection: If True, a fresh discrete collision detection is run before the contact point query. Pass False to skip if a recent sim step or pre-process has run a collision detection pass on the current state.
+    :param vertical_normal_error_threshold: The allowed error in normal alignment for a contact point to be considered "vertical" for this check. Functionally, if dot(contact normal, Y) <= threshold, the contact is ignored.
+
+    :return: a list of integer object_ids for the set of objects "ontop" of objectA.
```

Comment:
Will users ever need to check whether a specific `object_b` is on top of `object_a`, and get a simple bool as a result? A variant function may be cheaper if they need to do this comparison often.

Reply:
It could be a bit cheaper, but you'd still need to look through every contact point, so I don't think it would be much faster. I guess the advantage would the option for early abort.

Reply:
I'll leave this to a future PR. Early abort and a check against provided object_id should be easy enough to add in a small refactor if we want it.

Reply:
Sounds good.
My worry is that users would also have to find the desired object within the results - the search itself could be called in a nested loop, making this overall process inefficient.
I agree with you - let's see how this ends up being used.